### PR TITLE
ARROW-16613: [Python][Parquet] pyarrow.parquet.write_metadata with metadata_collector appears to be O(n^2)

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref Code:

<details>

```
from io import BytesIO

import pyarrow as pa
import pyarrow.parquet as pq
from contexttimer import Timer  # non standard lib (can be installed with pip)


def create_example_file_meta_data():
    data = {
        "str": pa.array(["a", "b", "c", "d"], type=pa.string()),
        "uint8": pa.array([1, 2, 3, 4], type=pa.uint8()),
        "int32": pa.array([0, -2147483638, 2147483637, 1], type=pa.int32()),
        "bool": pa.array([True, True, False, False], type=pa.bool_()),
    }
    table = pa.table(data)
    metadata_collector = []
    pq.write_table(table, BytesIO(), metadata_collector=metadata_collector)
    return table.schema, metadata_collector[0]

schema, meta = create_example_file_meta_data()
print("Created Example File")
metadata_collector = [meta] * 500
with Timer(prefix='1'):
    pq.write_metadata(schema, BytesIO(), metadata_collector=metadata_collector)

metadata_collector = [meta] * 1000
with Timer(prefix='2'):
    pq.write_metadata(schema, BytesIO(), metadata_collector=metadata_collector)

metadata_collector = [meta] * 2000
with Timer(prefix='3'):
    pq.write_metadata(schema, BytesIO(), metadata_collector=metadata_collector)

metadata_collector = [meta] * 4000
with Timer(prefix='4'):
    pq.write_metadata(schema, BytesIO(), metadata_collector=metadata_collector)
```

</details>

Before
```
Created Example File
1 took 0.615 seconds
2 took 2.446 seconds
3 took 9.813 seconds
4 took 40.237 seconds
```

After
```
Created Example File
1 took 0.009 seconds
2 took 0.018 seconds
3 took 0.036 seconds
4 took 0.072 seconds
```

TODO:
* [x] To actually overload Cython function than adding a new one.
* [x] Test to see that we are not messing anything up.

### CODE REVIEW

**Comment by kshitij12345** [comment]
cc: @AlenkaF

**Comment by AlenkaF** [comment]
Thanks @kshitij12345 for contributing!

Current work looks perfect. Interested to see the continuation of it, join of two cython functions into one to accept one or a list of `FileMetaData` objects and also the tests.

Could you change the name of the PR to `ARROW-16613: [Python][Parquet] pyarrow.parquet.write_metadata with metadata_collector appears to be O(n^2)` so it connects to the JIRA ticket you are working on?

And I will ask for the workflows to get started on this PR.

File: cpp/src/parquet/metadata.cc [hunk 1]
```diff
@@ -664,6 +664,34 @@ class FileMetaData::FileMetaDataImpl {
     }
   }
 
+  void AppendRowGroups(const std::vector<std::shared_ptr<FileMetaData>>& others) {
+    // Figure out the total num_groups and reserve vector accordinly.
```
Comment (kshitij12345) [inline cpp/src/parquet/metadata.cc:668]: accordinly -> accordingly

**Comment by AlenkaF** [comment]
Thank you for your work! This PR looks great to me.
@jorisvandenbossche @pitrou could you also have a look at it?

**Comment by kshitij12345** [comment]
Ah. That is neat :)
Closing in favour of #13265

**Pull request closed without merging**